### ISSUE TITLE
NodeSshKeyAdd does not handle updates gracefully

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#1102]
Originally reported of Google Code with ID 1046.

```
backend/NodeSshKeyAdd aborts the entire operation as soon as upgrading one node fails. It should handle this more gracefully.
```
Originally added on 2015-03-20 13:21:44 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Fixed with:

commit bacd27531526faa60e595b3b4a4498a35b258779
Author: Helga Velroyen <helgav@google.com>
Date:   Fri Apr 17 14:35:06 2015 +0200

    NodeSshRemoveKey: Add retries for updating the target node
    
    This patch adds retries to the part of 'NodeSshRemoveKey',
    where the target node itself is updated. As this is likely
    to fail if the node is for example already offline, we
    do not fail the entire operation, but emit a non-disruptive
    error which can be showing as feedback by the LU.
    
    Signed-off-by: Helga Velroyen <helgav@google.com>
    Reviewed-by: Hrvoje Ribicic <riba@google.com>


commit b54540a37bc97b8c6a1a358e8cfbff519692a2b0
Author: Helga Velroyen <helgav@google.com>
Date:   Fri Apr 17 12:45:03 2015 +0200

    RemoveNodeSshKey: use retries when updating other nodes
    
    When removing an SSH key, RemoveNodeSshKey so far did only
    try to contact each node once. This made the entire
    procedure not very resilient towards temporary network
    failures. This patch adds a couple of retries for contacting
    each node and only gives up after those are exhausted.
    
    Signed-off-by: Helga Velroyen <helgav@google.com>
    Reviewed-by: Hrvoje Ribicic <riba@google.com>

commit 6ad504bd7696e49e9f68d9e627112a5ad866dc94
Author: Helga Velroyen <helgav@google.com>
Date:   Fri Apr 17 11:47:54 2015 +0200

    AddNodeSshKey: retries for all non-master nodes
    
    This patch introduces retries for updating non-master nodes
    when a new SSH key is added to the cluster. If an SSH update
    fails even after the maximum number of retries is exhausted,
    this does not raise an exception, but an error message is
    added to the result of the method, which can be displayed
    in the LU operations so that the admin sees which nodes
    were the problem and need further attention.
    
    Signed-off-by: Helga Velroyen <helgav@google.com>
    Reviewed-by: Hrvoje Ribicic <riba@google.com>

commit 1a30e46a910d772058d081ad2f155da2d2eee81a
Author: Helga Velroyen <helgav@google.com>
Date:   Thu Apr 16 16:55:16 2015 +0200

    AddNodeSshKey: retry when target node not reachable
    
    So far, if a SSH key of a node is added to the cluster
    and updating the SSH setup of the node itself fails,
    the entire operation fails quite disgraceful. This
    patch introduces a retry mechanism and a proper cleanup
    in case of a failure.
    
    Note that there are more places where retries would
    be a good idea. They'll be added in other patches.
    
    Signed-off-by: Helga Velroyen <helgav@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 49f49f81148ffcde660eee58ccfdafb2010e2329
Author: Helga Velroyen <helgav@google.com>
Date:   Thu Apr 16 15:02:47 2015 +0200

    Consider offline nodes when removing SSH keys
    
    Like the previous patch, also NodeSshRemoveKey did not
    consider offline nodes so far. This patch fixes it.
    
    This fixes issue 1047.
    
    Signed-off-by: Helga Velroyen <helgav@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit d95ba4afd730c87198cfd17ae32e5211a2723d64
Author: Helga Velroyen <helgav@google.com>
Date:   Thu Apr 16 15:02:08 2015 +0200

    Consider offline nodes in NodeSshKeyAdd
    
    So far the backend function NodeSshKeyAdd naively assumed
    that all nodes are online and reachable via SSH. This
    patch adds offline status information to the function's
    signature to respect the offline status when updating
    SSH keys.
    
    Signed-off-by: Helga Velroyen <helgav@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>
```
Originally added on 2015-04-23 15:02:04 +0000 UTC.

Mentioned Issue with GoogleCode ID 1047 has been migrated to GitHub with ID #1103.
Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Released as part of 2.14.0
```
Originally added on 2015-06-02 10:37:03 +0000 UTC.

Changed State: Released
